pkg/models/job_history.go

```diff
@@ -41,13 +42,15 @@ type StateChange[StateType any] struct {
 // {Job,Event}State fields will only be present if the Type field is of
 // the matching type.
 type JobHistory struct {
-	Type           JobHistoryType                   `json:"Type"`
-	JobID          string                           `json:"JobID"`
-	NodeID         string                           `json:"NodeID,omitempty"`
-	ExecutionID    string                           `json:"ExecutionID,omitempty"`
-	JobState       *StateChange[JobStateType]       `json:"JobState,omitempty"`
-	ExecutionState *StateChange[ExecutionStateType] `json:"ExecutionState,omitempty"`
-	NewRevision    uint64                           `json:"NewRevision"`
-	Comment        string                           `json:"Comment,omitempty"`
-	Time           time.Time                        `json:"Time"`
+	Type             JobHistoryType                   `json:"Type"`
+	JobID            string                           `json:"JobID"`
+	NodeID           string                           `json:"NodeID,omitempty"`
+	ExecutionID      string                           `json:"ExecutionID,omitempty"`
+	EvaluationID     string                           `json:"EvaluationID,omitempty"`
+	EvaluationStatus string                           `json:"EvaluationStatus,omitempty"`
+	JobState         *StateChange[JobStateType]       `json:"JobState,omitempty"`
+	ExecutionState   *StateChange[ExecutionStateType] `json:"ExecutionState,omitempty"`
+	NewRevision      uint64                           `json:"NewRevision"`
+	Comment          string                           `json:"Comment,omitempty"`
+	Time             time.Time                        `json:"Time"`
```

Comment:
I am not particularity keen on extending this structure to contain more information in its present form. The cognitive overhead of remembering which values are set based on the type high. Ideally we will move towards something like:
```go
type Event struct {
	Type    EventType
	JobID   string
	Time    time.Time
	Payload []byte
}
```
in future implementations where `Type` dictates the destination `Payload` decodes into. But I suspect that's a topic for a later date.

Reply:
Yeah I agree this is not great, but I also agree with Ross that it is probably better to wait and evaluate better ways to keep track of such events when there is a pressing need.

Reply:
Lets take a step back and understand what we have today and what we are missing to be able to show job progress to the user. 

Now the job history and execution history events are already capturing when the job is created, execution progress, and completion. Adding evaluation creation and update events will show as duplicate events in the history as most evaluations are created after a job/execution state is updated.

What job and execution events are not capturing are:
- Reasons behind the creation or stopping of executions. This might be mitigated by enriching the history event and add more information to the comment field.
- Events that did not result in job or execution state update, which won't show in the history. An example is when scheduling fails and we decide to retry later. We don't have this feature yet and is part of this PR https://github.com/bacalhau-project/bacalhau/pull/3135
- Events that are specific to the job request so that when we enable job updates, you only get events about the job update and not old events from the job creation. Again we don't have this feature yet if you don't want to worry about it now, and the mitigation would be to add the EvalID in the job and execution history events instead of adding separate evaluation events

So it sounds like what you need to do to improve the job progress visibility, is to enrich the existing job and execution history events with additional metadata or better comment messages. I don't think adding events for evaluation state updates will help much

Reply:
Let's look back at the issue https://github.com/bacalhau-project/expanso-planning/issues/389 this work is attempting to address, which states:

1. The proposal is to use evaluations to track the job progress
2. Expose APIs to query evaluations.
3. enrich the job submission command.
4. Chain evaluations so that we can always reach child evaluations from a parent one.
5. Enrich evaluations with metadata that are useful to understand the scheduler’s decisions. These metadata can be a list of arbitrary messages, such as number of discovered nodes, filtered nodes.
6. We are mainly interested in the evaluation that was created when the job was submitted, along with its executions and any evaluation created while trying to bring those executions to a running or completed state
 
This PR addresses points 1 and 2 - 3 will come in a follow on to this work as will 5.

We have come to an understanding 4 isn't practically feasible given our current approach for creating evaluations -  although https://github.com/bacalhau-project/bacalhau/pull/3500 moves us closer to doing so via post-processing.

However, what I am understanding here is that we want something slightly different. 

-------------------

> What job and execution events are not capturing are:
> - Reasons behind the creation or stopping of executions. This might be mitigated by enriching the history event and add more information to the comment field.

I _think_ this is what the `TriggeredBy` field of an evaluation is for, I have asked about it here: https://github.com/bacalhau-project/bacalhau/pull/3499#discussion_r1496699255

> Events that did not result in job or execution state update, which won't show in the history.

Yea, these would be great to have! But also sounds like a new event type - separate from jobs, executions, or evaluations. 
One way to obtain this information is by tracking the conditions and why they weren't met via the update request sent to the job store. But since this change is already landing in https://github.com/bacalhau-project/bacalhau/pull/3135 I guess it's probably better to wait and evaluate better ways to keep track of such events when the need arises.

> Events that are specific to the job request so that when we enable job updates, you only get events about the job update and not old events from the job creation

I'm not sure I understand this bit. We already update jobs in the job store. Unless you mean allowing users to update their jobs?

> the mitigation would be to add the EvalID in the job and execution history events instead of adding separate evaluation events

Please inform me of how you would like this work done:
- Job and execution history events all have comments associated with them. Evaluations also have comments. Should these fields be merged in the corresponding job or execution events comment? 
- You have alluded to collecting other information from the evaluation in your commends - such as `Status` and `TriggeredBy`. How can this information be represented in the job or execution event?

Reply:
My apologies if the goals were not clear. The goal of the issue you are referring to is to improve the job progress visibility when users call `bacalhau docker run` or `bacalhau job run`. Using evaluations was just a proposal, but you found some challenges with that route as you pointed out in the issue, mainly with tracking the lineage. You have actually suggested a good alternatives where the client listens for job store events instead, but needed to be aware of the jobID before submission. 
We discussed in our meeting that you are free to use any other source to achieve better job progress visibility, and that enriching job history along with pagination would work similar to job store events, but are also persisted and can be queried after the fact.

> I think this is what the TriggeredBy field of an evaluation is for, I have asked about it here: https://github.com/bacalhau-project/bacalhau/pull/3499#discussion_r1496699255

Good point, but I don't think this will work as evaluations can be deduped by the evaluation broker where only one evaluation will be processed while others cancelled. This makes sense because the scheduler only compares the desired job state with the actual observed state to make scheduling decisions, and doesn't care about the source or trigger of the evaluation

> Yea, these would be great to have! But also sounds like a new event type - separate from jobs, executions, or evaluations.

Not necessary. To me it sounds like a job level event as it did not touch a specific execution, and the whole job will fail eventually if we didin't schedule the job. If we keep things simple, anything event that is not related to specific execution is most likely a job level event. 

One way for us to add a history event when scheduling is delayed is to actually update the job state from `running` to `pending` as an example. Another way is to expose an API to add history events directly by the scheduler and not only after `state` updates. Though as you pointed out, we don't need to worry about this for now and should be handled with/after https://github.com/bacalhau-project/bacalhau/pull/3135

> I'm not sure I understand this bit. We already update jobs in the job store. Unless you mean allowing users to update their jobs?

Correct, when the user update their job and want to see progress for the update request and not older history events

> Job and execution history events all have comments associated with them. Evaluations also have comments. Should these fields be merged in the corresponding job or execution events comment?

Maybe. Not all job/execution events have a comment. We just need to evaluate each history event to come up with the right solution on what to include in the comment field. 

> You have alluded to collecting other information from the evaluation in your commends - such as Status and TriggeredBy. How can this information be represented in the job or execution event?

Similar to my answer above

I hope this helps, and let me know if I can clarify anything else